ark/mibi/qc_comp.py

```diff
@@ -1,14 +1,169 @@
+    if len(run_info['results']) == 0:
+        raise ValueError('No data found for run_name %s and run_label %s' % (run_name, run_label))
+
+    # extract the name of the FOVs and their associated internal IDs
+    run_fov_names = [img['number'] for img in run_info['results'][0]['imageset']['images']]
+    run_fov_ids = [img['id'] for img in run_info['results'][0]['imageset']['images']]
+
+    # if fovs is None, ensure all of the fovs in run_fov_names_ids are chosen
+    if fovs is None:
+        fovs = run_fov_names
+
+    # ensure all of the fovs are valid (important if the user explicitly specifies fovs)
+    misc_utils.verify_in_list(
+        provided_fovs=fovs,
+        mibitracker_run_fovs=run_fov_names
+    )
+
+    # extract the name of all the channels
+    # NOTE: this set of channels will be the same across all FOVs in the run
+    run_channels = run_info['results'][0]['imageset']['images'][0]['pngs']
+
+    # if channels is None, ensure all of the channels in run_channels are chosen
+    if channels is None:
+        channels = run_channels
+
+    # ensure all of the channels are valid (important if the user explicitly specifies channels)
+    misc_utils.verify_in_list(
+        provided_chans=channels,
+        mibitracker_run_chans=run_channels
+    )
+
+    # if the desired tiff_dir exists, remove it
+    if os.path.exists(os.path.join(base_dir, tiff_dir)):
+        rmtree(os.path.join(base_dir, tiff_dir))
+
+    # make the image directory
+    os.mkdir(os.path.join(base_dir, tiff_dir))
+
+    # ensure sub_folder gets set to "" if img_sub_folder is None (for os.path.join convenience)
+    if not img_sub_folder:
+        img_sub_folder = ""
+
+    # define the run order list to return
+    run_order = []
+
+    # iterate over each FOV of the run
+    for img in run_info['results'][0]['imageset']['images']:
+        # if the image fov name is not specified, move on
+        if img['number'] not in fovs:
+            continue
+
+        print("Creating data for fov %s" % img['number'])
+
+        # make the fov directory
+        os.mkdir(os.path.join(base_dir, tiff_dir, img['number']))
+
+        # make the img_sub_folder inside the fov directory if specified
+        if len(img_sub_folder) > 0:
+            os.mkdir(os.path.join(base_dir, tiff_dir, img['number'], img_sub_folder))
+
+        # iterate over each provided channel
+        for chan in channels:
+            # extract the channel data from MIBItracker as a numpy array
+            chan_data = mr.get_channel_data(img['id'], chan)
+
+            # define the name of the channel file
+            chan_file = '%s.tiff' % chan
+
+            # write the data to a .tiff file in the FOV directory structure
+            imsave(
+                os.path.join(base_dir, tiff_dir, img['number'], img_sub_folder, chan_file),
+                chan_data
+            )
+
+        # append the run name and run id to the list
+        run_order.append((img['number'], img['id']))
```

Comment:
was it tested if iterating over the `run_info` will always yield fovs in the run order?

Reply:
The sorting scheme in the notebook accounts for the fact `fovs` will not necessarily be retrieved in the run order. This is done by keeping track of the image ID retrieved for each image, which are sorted in run order. By using this image ID as a key for sorting, the actual FOV run order can be maintained.